website/main/content/blog/2025-06-30-sector-expiration.md

```diff
@@ -0,0 +1,213 @@
+
+### Overview
+
+Before I start with the description of the protocol, let me remind you what are the core properties
+that we wanted the protocol to have:
+
+1.  **Unique Plot IDs:** `plot_id` is unique and stable. This is key for the protocol, as this ID
+    will be used to allocate space into shards.
+2.  **No Bias through `history_size`:** Farmers can't game shard allocation by choosing among any
+    `history_size`, or by creating sectors with the same ID committed to different history size.
+    This prevents an attack where the same plot ID could be used to create sectors with different
+    `history_size` values, which would allow farmers to manipulate shard allocation by assigning to
+    a plot the most convenient sector at each time.
+3.  **Load-Balanced Archiving (New History):** As new history arrives, it gets stored across
+    farmers/plots without large re-plotting events.
+4.  **Detect Expiration and Validity:** Plots, sectors and blocks include all the information needed
+    to verify their validity, plot assignment to a shard, sector belonging to the plot, and
+    expiration.
+5.  **No Excessive Re-plotting:** Keep re-plotting burden similar to original Subspace.
+
+### Committing sectors to history ranges
+
+The core idea for sector expiration is the following: instead of letting farmers commit their
+sectors to any history size, they are bound to commit them to a range of history sizes determined by
+the plot the sector is linked to. This way, we limit the range of history sizes that a sector can be
+committed to, and hence the parallel sectors with the same `plot_id` that and different committed
+history sizes that can be created. This is implemented as a "self-evolving plot commitment range"
+that grows as the blockchain grows on top of the existing mechanism for sector expiration. Thus, we
+maintain all the properties of the original Subspace protocol, while adapting it to the needs of the
+sharded version. This allows us to also learn an apply the heuristics from the live Subspace
+network.
+
+The high-level operation of the protocol is as follows:
+
+- When creating a new plot, farmers need to choose a random offset (essentially a nonce) for their
+  plot. This offset determines the specific ID of the plot and will influence the history range that
+  the plot's sectors can commit to. As a reminder, plots identities are inferred from the public key
+  hash of the farmer, and the random offset they choose:
+
+```rust
+plot_id = hash(public_key_hash || offset)
+```
+
+- When creating new sectors for a plot, farmers will choose any valid history size within the plot's
+  current range. Piece selection is based on the sector's specific history size, and expiration
+  follows the same logic as the original Subspace protocol. Finally, the sector ID is derived from
+  the plot ID, sector index, and committed history size:
+
+```rust
+sector_id = keyed_hash(plot_id, sector_index || committed_history_size)
```

Comment:
They commit to a range, not a specific history size

Reply:
Does it? I actually was thinking that the sector would commit to a specific history size chosen by the farmer within the range so no changes are required to the expiration logic. We can commit to a history range too, in which case the beginning or the end of the range will determine the history size used for the expiration logic. 

No strong opinions, happy to introduce whatever you think is best (I don't think it changes much)

Reply:
If they can choose multiple history sizes within a range, they can create multiple valid sectors for the same `sector_index`. The whole point of the history range was to make them commit to a range instead, resulting in a single `sector_id` per sector range instead.